test/fw/ptl/lib/pbs_testlib.py

```diff
@@ -6568,6 +6575,11 @@ def manager(self, cmd, obj_type, attrib=None, id=None, extend=None,
                     bsl = self.utils.convert_to_dictlist(ret['out'], attrib,
                                                          mergelines=True)
                     self.update_attributes(obj_type, bsl)
+                elif cmd == MGR_CMD_DELETE:
+                    if id in self.schedulers.keys():
+                        del self.schedulers[id]
+                elif obj_type == SCHED:
+                    self.manager(MGR_CMD_LIST, obj_type, id=id)
```

Comment:
@visheshh , 
why have you put "del self.schedulers[id]" at this place when there is a common place for deletion of such "id" for different PBS objects(at line 6628)

Also I do not see any use of this line(6582):
                elif obj_type == SCHED:
                    self.manager(MGR_CMD_LIST, obj_type, id=id)
why are u calling this?

Reply:
Hi @varunsonkar , 
qmgr list sched will query and update the latest information to schedulers dictionary , So that every time we will do any qmgr set , unset, create update there is no need for us to do qmgr list sched inside the test case to get the latest info and I do not want to do list if it is qmgr delete of the id , because that wont be available , So i am deleting it from schedulers dict .
I agree it will be nice to have it all at one place at line 6628 as you suggest .
If i move it there i need to add one more check before qmgr list that it is not a qmgr delete command.
What would you suggest ?

Reply:
@visheshh ,
For this specific purpose we pass "expect=True" in "manager" to validate set/unset/create.
Anyways MGR_CMD_LIST will not validate anything.
Please try using expect.

Reply:
Hi @varunsonkar , 
I thought through and tested as you say , expect=True validates whatever we set is set and whatever we unset is removed .
I am not trying to validate anything using MGR_CMD_LIST , I am just trying to update schedulers dictionary for future reference , So that i save a lot of rework , 
For example:
1. If i set something invalid and test that the default value still holds good 
2. If i unset something in scheduler like sched_priv,sched_log,sched_port  or state It does not delete the entry which expect=True in unset checks for, but its set to a default value which is not checked by unset=True (We can enhance, again that will be specific to few attrs which is not good)
3.When you create a scheduler expect=true does not know to verify that there is sched_priv,sched_log,state is created and is set to some value .
4. When scheduler is started sched_port and few attrs are set , I am doing this in other place where you have asked the same question.

In all these cases , If i do not do a qmgr list here , I have to do each and every time where i run into above mentioned cases test cases .
I feel that even for future test case writers , It will be very useful that they do not call qmgr list every time they come into any of these 4 scenarios .
If they set/unset/create some scheduler attr , We will always be sure that Schedulers dict will be updated and they can query for their usage.

For an instance : If somebody set sched_log directory to new directory for sched1 and did not do expect=True , schedulers['sched1'].log_match will fail as the dictionary is not updated and it is looking into the old scheduler log directory

Let me know your opinions .

Reply:
@visheshh ,
At any step of set/create/unset you have to validate.
"expect" function does that.
Only case you might be worried is if unset reverts the attribute to default value, in that case you should not pass "expect=True" but call expect function in the next step exclusively to validate that the attribute has been reverted to default value.
As per the testing rule if you do set/unset/create , you will have to validate using expect.
Which will anyways update your attribute/object list/dictionary

Reply:
@varunsonkar ,
If i am only doing validation , I accept that using expect=True or calling expect function would suffice rather than doing a qmgr list and asserting them.
The main objective here is to maintain data consistency in schedulers dictionary once something gets updated in PBS.
Lets consider another example in log_match.
My current directory is PBS_HOME/sched_new_logs_sc1 .
I want to unset this directory the default value and use log match on default directory .
Then when i do a MGR_CMD_UNSET 
PBS reverted it back to PBS_HOME/sched_logs_sc1 
But schedulers['sc1'].log_match will fail because PTL Schedulers dict value still has old value .
To make it work I need to call expect function which will do qmgr list .

To maintain data consistency between PTL Framework and PBS , I do not see why we should wait from an expect function call or expect=True from a test writer ?

As a test writer i will always want framework to handle consistency of set data on PBS and PTL data to be same . In order to maintain it we need to update schedulers dictonary attribute.
As qmgr list already updated the attributes and helps us maintain data consistency , I am calling it .

Reply:
@visheshh ,
I have two points here, which you should take into consideration when changing the framework:

1) We should not make framework changes for case specific. Up till  now before your changes the framework was not differentiating  between pbs objects for the use cases similar to the one you have mentioned. If you still feel that this is the way to go ahead. I will recommend to make it consistent with all the pbs objects not just sched.
Also, discuss this proposal with PTL steering committee  and get your feedback on this change.

2) Also you should not call manager function within the manager function. Please use something like this which we do to update the attributes:

                    bsl = self.utils.convert_to_dictlist(ret['out'], attrib,
                                                         mergelines=True)
                    self.update_attributes(obj_type, bsl)

Reply:
@varunsonkar , 
1. I will reconfirm from maintainers and check if we should do this for all object types .
2. Here update attributes uses ret['out'] from qmgr list , which we do not get when we set/create/unset we only get return code . Even if i get this from inputs to set command.
For unset and few attributes of create we should again run list command to get the attribute value .